### PR TITLE
rsx/spu: Performance optimizations and other improvements

### PR DESCRIPTION
**Pull request by kd-11**
Highlights

1. RSX
 - Avoid aggressive resource create/delete cycles when using vulkan. Moderate speedup in some games and fixes flickering in some cases
 - Fix a vulkan crash about undeclared fog_c. Also fixes default parameter initialization that should fix some games that have broken graphics with vulkan but work fine with openGL
- Fix an rsx crash when a null address is provided for the fragment shader location
- Improved multithreaded vertex processing so that the penalty is much lower if there are no resources available. Its also tunable now allowing the threshold to be set

2. SPU
- Avoid aggressively locking the asmjit db and make compilation step multithreaded. It takes a much shorter time to compile a function than it does to wait on a lock especially with non-spurs-type kernels
- Add loop condition detection by triggering an OS scheduler update on RdDec - this function can only be sensibly used as part of a loop. [Disabled by default for now]
- Add concurrent execution analysis. This only affects spurs-type kernels where multiple threads are executing the exact same code at the exact same time. Introduces a small delay to racing threads so that they are effectively desynchronized whenever they enter a sensitive function. [Disabled by default]

3. Other
- CMake fix to not double-compile qrc_resources.cpp as it is a tmp generated file. Also adds it to list of files to clean

TODO before merge:
- [x] Reduce watchdog memory footprint by using the segment address as a base + 256K
- [x] Investigate tales of vesperia regression

### CODE REVIEW

**Comment by zminhquanz** [comment]
You need to fix random timing of the lower left corner of the note is not full of black texture error, and some 2d material can not be read on Project Diva F

**Comment by AniLeo** [comment]
Is that related to this PR at all? If so, what song?

**Comment by Lemiru** [comment]
For Deception IV: The Nightmare Princess this PR fixed this crash:
```
E {rsx::thread} RSX: ERROR: 0:65: 'tc9' : undeclared identifier 
ERROR: 0:65: '' : compilation terminated 
ERROR: 2 compilation errors.  No code generated.


E {rsx::thread} RSX: 
F {rsx::thread} class std::runtime_error thrown: Failed to compile vertex shader
(in file C:\rpcs3\rpcs3\Emu\RSX\VK\VKVertexProgram.cpp:401)
```

**Comment by Xcedf** [comment]
Assassin's Creed and Prince of Persia now go ingame


but Resident Evil Revelations now has flickering issues under OGL

**Comment by Xcedf** [comment]
Resident Evil  Revelations issue i've reported now fixed on latest rev
Thanks

**Comment by Xcedf** [comment]
After the last commit noticed that performance gains from SPU threads were killed, but reverting it gave me nothing, strange but reverting d7a9643 returns the speed, the problem didn't even existed before 86bd6b6

**Comment by kd-11** [comment]
I know it's faster without the reader lock but I cant guarantee non msvc compilers like gcc and clang will not crash so I added it back for now. A better solution will be more stable. If building for yourself on windows you can comment out the reader lock

**Comment by kd-11** [comment]
Also spu threads does not work the way you think. Its either better performance with it at 1 or 2 or it works better disabled. If you're using 4 threads you will likely only benefit from loop detection

**Comment by Xcedf** [comment]
Thanks for the info

File: rpcs3/Emu/Cell/SPUAnalyser.cpp [hunk 1]
```diff
@@ -12,9 +12,24 @@ std::shared_ptr<spu_function_t> SPUDatabase::find(const be_t<u32>* data, u64 key
...
-		if (LIKELY(func->size <= max_size) && std::memcmp(func->data.data(), data, func->size) == 0)
+		//if (LIKELY(func->size <= max_size) && std::memcmp(func->data.data(), data, func->size) == 0)
 		{
-			return func;
+			const auto dwords = (u32*)(func->data.data());
+			const auto cast_data = (u32*)(data);
+			const auto limit = std::min(max_size, func->size) >> 2;
+
+			bool failed = false;
+			for (u32 dword = 0; dword < limit; dword++)
```
Comment (danilaml) [inline rpcs3/Emu/Cell/SPUAnalyser.cpp]: Are you sure this is better than simple `memcmp`? AFAIK it's usually vectorized as well.

Comment (kd-11) [inline rpcs3/Emu/Cell/SPUAnalyser.cpp]: Nope. I'm undoing most of this commit actually. The compiled blocks are actually so small that its not worth it

**Comment by kd-11** [comment]
@Xcedf Performance issues due to the locks are resolved now in a much cleaner way. You may retest.

**Comment by Xcedf** [comment]
**@kd-11** Confirm. Performance problems resolved, things even slightly faster now

**Comment by Xcedf** [comment]
not tested T6 for long time
it now hits fullspeed sometimes
avg fps 54-58 on this branch on many areas and characters
game is pretty much Playable now

**Comment by zminhquanz** [comment]
It's improve perfomance on Project Diva F too

**Pull request merged**